lib/core/macros.nim

```diff
@@ -1344,6 +1348,76 @@ proc copy*(node: NimNode): NimNode {.compileTime.} =
   ## An alias for copyNimTree().
   return node.copyNimTree()
 
+type GenAstOpt* = enum
+  kDirtyTemplate,
+    # When set, uses a dirty template in implementation of `genAst`. This
+    # is occasionally useful as workaround for issues such as #8220, see
+    # `strformat limitations <strformat.html#limitations>`_ for details.
+    # Default is unset, to avoid hijacking of uncaptured local symbols by
+    # symbols in caller scope.
+  kNoNewLit,
+    # don't call call newLit automatically in `genAst` capture parameters
+
+macro genAstOpt*(options: static set[GenAstOpt], args: varargs[untyped]): untyped =
+  ## Accepts a list of captured variables `a=b` or `a` and a block and returns the
+  ## AST that represents it. Local `{.inject.}` symbols (e.g. procs) are captured
+  ## unless `kDirtyTemplate in options`.
+  runnableExamples:
+    macro fun(a: bool, b: static bool): untyped =
+      let c = false # doesn't override parameter `c`
+      var d = 11 # var => gensym'd
+      proc localFun(): auto = 12 # proc => inject'd
+      genAst(a, b, c = true):
+        # echo d # not captured => gives `var not init`
+        (a, b, c, localFun())
+    doAssert fun(true, false) == (true, false, true, 12)
+
+  let params = newTree(nnkFormalParams, newEmptyNode())
+  let pragmas =
+    if kDirtyTemplate in options:
+      nnkPragma.newTree(ident"dirty")
+    else:
+      newEmptyNode()
+
+  template newLitMaybe(a): untyped =
+    when type(a) is NimNode: a
+    elif compiles(newLit(a)): newLit(a)
```

Comment:
Depending on ``compiles`` is very harmful. Sometimes the compiler becomes stricter and prints error messages. ``compilers`` eats them and secretly changes behavior.

Reply:
found a better way without `compiles`;
(makes me think would be nice to have a `compiles(foo(args), sigcheckOnly)` which would just return whether a proper overload can be found, without compiling body; this would also be more efficient and resolve a number of issues with `compiles`)

Reply:
we actually think about something like that internally. Something like ``overloadExists(foo(args))``.

Reply:
=> moved that discussion to https://github.com/nim-lang/RFCs/issues/164